### PR TITLE
Use the same  prometheus registry for both oc and otel exporters

### PR DESCRIPTION
**Pull request by paivagustavo**
**Description:** Use the same `prometheus.Registry` for both OC and OTel SDK exporters.

**Link to tracking Issue:** #5687

**Testing:** Added unit test and did some manual tests with and without the featuregate to ensure that nothing is broken.

**Documentation:** The observed difference between both exporters is how they export Resource Attributes. OpenCensus appends resource attributes to all metrics while [OpenTelemetry advocates](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#resource-attributes-1) for the use of an additional `target_info` metric which contains all resource attributes and other metrics are only appended with `job` and `instance` which can be derived from the `service.name`, `service.namespace`, and `service.instance.id attributes`.

OTel-Go doesn't implement the `target_info` yet, but I've added some notes and TODOs to track those.

### CODE REVIEW

File: service/telemetry.go [hunk 1]
```diff
@@ -205,12 +214,13 @@ func (tel *telemetryInitializer) initOpenCensus(cfg telemetry.Config, telAttrs m
...
 	if err := view.Register(views...); err != nil {
-		return nil, err
+		return err
 	}
 
 	// Until we can use a generic metrics exporter, default to Prometheus.
 	opts := ocprom.Options{
-		Namespace: "otelcol",
+		Namespace:  "otelcol",
+		Registerer: promRegistry,
```
Comment (bogdandrutu) [inline service/telemetry.go:223]: Not sure if it is not better to pass the full `Registry` here. Have not checked the code to see if it has any usage of the Gatherer/Registry. I think it is safer to pass the `Registry: registry`

Comment (bogdandrutu) [reply service/telemetry.go:223]: And even maybe use for the moment the Handler from the OC, since I am not 100% if there is no side effect there..

Comment (paivagustavo) [reply service/telemetry.go:223]: The `Gatherer` is just used if we want to use the exporter itself as the handler.

Comment (bogdandrutu) [reply service/telemetry.go:223]: :)) famous last words. I am not confident, so still prefer to send the entire Registry to OC and use the Handler form there, since OC is the current state that guarantees me that no new code is used. Look at the PR I created.

Comment (paivagustavo) [reply service/telemetry.go:223]: I did check and couldn't side effects for not using the gatherer. I can double check to make sure, but since we would want to get rid of the oc exporter in the future, we would to create our handler anyway.

Comment (paivagustavo) [reply service/telemetry.go:223]: Sure! I understand your point :)

**Review by bogdandrutu — APPROVED** [review]

**Comment by bogdandrutu** [comment]
Please update descriptions, and add some unit tests

**Comment by paivagustavo** [comment]
@bogdandrutu I've added some unit tests, not sure if that was the best place, open for suggestions if you have any. 

Also documented on the tests the differences between how both SDK exports Resource attributes. I'm not sure what is the best way for us to document this public-wise, since it can be considered as a breaking change for some.

**Comment by bogdandrutu** [comment]
I would prefer proper unittests, I know we have some kind of "integration" tests there, but they are actually complicated and hard to follow. I would rather remove them and have proper unit-tests.

File: service/telemetry.go [hunk 2]
```diff
@@ -84,22 +83,28 @@ func (tel *telemetryInitializer) init(buildInfo component.BuildInfo, logger *zap
...
-	if cfg.Metrics.Level == configtelemetry.LevelNone || cfg.Metrics.Address == "" {
-		logger.Info(
-			"Skipping telemetry setup.",
-			zap.String(zapKeyTelemetryAddress, cfg.Metrics.Address),
-			zap.String(zapKeyTelemetryLevel, cfg.Metrics.Level.String()),
-		)
-		return nil
-	}
-
+func (tel *telemetryInitializer) initOnce(buildInfo component.BuildInfo, logger *zap.Logger, cfg telemetry.Config) error {
```
Comment (paivagustavo) [inline service/telemetry.go:107]: I've moved some of the code from `initOnce` in order to avoid starting a http server for the prometheus handler, since that is already captured by the tests in `collector_test.go`

File: service/telemetry_test.go [hunk 3]
```diff
@@ -67,3 +87,133 @@ func TestBuildTelAttrs(t *testing.T) {
...
+				require.NoError(t, tel.shutdown())
+			}()
+
+			v := createTestMetrics(t, tel.mp)
+			defer func() {
+				view.Unregister(v)
+			}()
+
+			// Wait some time because OpenCensus can take sometime to set up and collect a measure.
+			time.Sleep(10 * time.Millisecond)
```
Comment (paivagustavo) [inline service/telemetry_test.go:162]: I had to add this casual wait because sometimes we're not getting OpenCensus metrics. I ran this a few times with `-test.count 1000` and it would just randomly fail. Adding a millisecond wait was enough for me to stop seeing it, but just for safety I bumped it to 10 milliseconds.

I suspect that this does not affect the existing integration simply because they do more work before starting to test.

Comment (bogdandrutu) [reply service/telemetry_test.go:162]: I think if you call `view.RetrieveData(ocPrefix + counterName)` this will work

File: service/telemetry_test.go [hunk 4]
```diff
@@ -67,3 +87,133 @@ func TestBuildTelAttrs(t *testing.T) {
...
+	}
+}
+
+func createTestMetrics(t *testing.T, mp metric.MeterProvider) *view.View {
+	// Creates a OTel Go counter
+	counter, err := mp.Meter("collector_test").SyncInt64().Counter(otelPrefix + counterName)
+	require.NoError(t, err)
+	counter.Add(context.Background(), 13)
+
+	// Creates a OpenCensus counter
```
Comment (bogdandrutu) [inline service/telemetry_test.go:191]: This is not an opencensus "counter". This is an opencensus measure + sum aggregation view :)

File: service/telemetry_test.go (hunk 3 above)
Comment (paivagustavo) [reply service/telemetry_test.go]: That did the trick, thanks @bogdandrutu!

File: service/telemetry_test.go [hunk 5]
```diff
@@ -67,3 +86,133 @@ func TestBuildTelAttrs(t *testing.T) {
...
+		Description: ocCounter.Description(),
+		Measure:     ocCounter,
+		Aggregation: view.Sum(),
+	}
+	err = view.Register(v)
+	require.NoError(t, err)
+
+	stats.Record(context.Background(), stats.Int64(ocPrefix+counterName, counterName, stats.UnitDimensionless).M(13))
+
+	_, _ = view.RetrieveData(ocPrefix + counterName)
```
Comment (bogdandrutu) [inline service/telemetry_test.go:200]:
```suggestion
	// Forces a flush for the view data.
	_, _ = view.RetrieveData(ocPrefix + counterName)
```

**Review by bogdandrutu — APPROVED** [review]
